smartapp-core/src/main/java/com/smartthings/sdk/smartapp/core/service/TokenRefreshServiceImpl.java

```diff
@@ -0,0 +1,110 @@
+package com.smartthings.sdk.smartapp.core.service;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.nio.charset.StandardCharsets;
+import java.time.Clock;
+import java.time.Instant;
+import java.util.Arrays;
+import java.util.Base64;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.apache.http.NameValuePair;
+import org.apache.http.client.entity.UrlEncodedFormEntity;
+import org.apache.http.client.methods.CloseableHttpResponse;
+import org.apache.http.client.methods.HttpPost;
+import org.apache.http.entity.ContentType;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.message.BasicNameValuePair;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class TokenRefreshServiceImpl implements TokenRefreshService {
+    private final Logger log = LoggerFactory.getLogger(this.getClass());
+
+    private final String clientId;
+    private final String clientSecret;
+    private final String tokenRefreshUrl;
+
+    private final CloseableHttpClient httpClient;
+    private final Clock clock;
+
+    private static final ObjectMapper mapper = new ObjectMapper();
+
+    public TokenRefreshServiceImpl(String clientId, String clientSecret,
+            CloseableHttpClient httpClient) {
+        this.clientId = clientId;
+        this.clientSecret = clientSecret;
+        tokenRefreshUrl = "https://auth-global.api.smartthings.com/oauth/token";
```

Comment:
This URL has to be overridable for when we're using against preprod environments.